libr/anal/fcn.c

```diff
@@ -707,10 +710,85 @@ static bool is_delta_pointer_table(RAnal *anal, ut64 addr, ut64 ptr, ut64 *jmp_a
 			return false;
 		}
 	}
-	*jmp_addr = aop.addr;
 	return true;
 }
 
+static bool try_get_jmptbl_info(RAnal *anal, RAnalFunction *fcn, ut64 addr, RAnalBlock *my_bb, ut64 *table_size, ut64 *default_case){
+	bool isValid = false;
+	int i;
+	RListIter *iter;
+	RAnalBlock *tmp_bb, *prev_bb;
+	prev_bb = 0;
+	if (!fcn->bbs){
+		return false;
+	}
+
+	// search for the predecessor bb
+	r_list_foreach (fcn->bbs, iter, tmp_bb) {
+		if (tmp_bb->jump == my_bb->addr || tmp_bb->fail == my_bb->addr) {
+			prev_bb = tmp_bb;
+			break;
+		}
+	}
+	// predecessor must be a conditional jump
+	if (!prev_bb || !prev_bb->jump || !prev_bb->fail){
+		return false;
+	}
+
+	// default case is the jump target of the unconditional jump
+	*default_case = prev_bb->jump == my_bb->addr ? prev_bb->fail : prev_bb->jump;
+
+	RAnalOp tmp_aop = {0};
+	ut8 *bb_buf = malloc(prev_bb->size);
+	// search for a cmp register with a resonable size
+	anal->iob.read_at (anal->iob.io, prev_bb->addr, (ut8 *) bb_buf, prev_bb->size);
+	isValid = false;
+
+	for (i = 0; i < prev_bb->op_pos_size; i++) {
+	//for (int i = prev_bb->op_pos_size - 1; i >= 0; i--) {
+		ut64 addr = prev_bb->addr + prev_bb->op_pos[i];
+		int len = r_anal_op (anal, &tmp_aop, addr, bb_buf + prev_bb->op_pos[i], prev_bb->size - prev_bb->op_pos[i], R_ANAL_OP_MASK_ALL);
+
+		if (tmp_aop.type != R_ANAL_OP_TYPE_CMP) {
+			continue;
+		}
+		// get the value of the cmp
+		// for operands in op, check if type is immediate and val is sane
+		// TODO: How? opex?
```

Comment:
This one is tricky. I use RAnalOp.src/dst, but it largely not supported in radare2. This is a thing that needs to be done one way or another.

Reply:
I have tried using src/dst but they were always empty.
The workaround below seems to work pretty well so far. It handles register compares, as well as memory dereferences, but I wouldn't be surprised if there is a case where it breaks

Reply:
the src/dst thing is imho a bad thing to depend on, not all backends fill that and as long as it touches the heap it becomes pretty slow to generate/free

Reply:
@radare but what is the alternative to that? I have the same problem with struct offset recognition, where any of the opcodes can be a value or an offset?